cardano-cli/test/Test/CLI/TextEnvelope/Golden/Genesis/Create.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE OverloadedStrings    #-}
+
+module Test.CLI.TextEnvelope.Golden.Genesis.Create
+  ( golden_genesisCreate
+  ) where
+
+import Cardano.Prelude hiding (to)
+
+import Data.Maybe
+  ( fromJust
```

Comment:
If `fromJust` fails does it give a good error message about where the failure is coming from?

Reply:
You're right.  The user experience is terrible.

Reply:
How about this?

```
  ✗ golden_genesisCreate failed at test/Test/CLI/TextEnvelope/Golden/Genesis/Create.hs:62:22
    after 1 test.
  
       ┏━━ test/Test/CLI/TextEnvelope/Golden/Genesis/Create.hs ━━━
    60 ┃ justResult :: (Show a, HasCallStack) => Maybe a -> H.PropertyT IO a
    61 ┃ justResult (Just a) = H.forAll $ pure a
       ┃ │ 10000000
       ┃ │ "2020-07-10T07:44:17Z"
       ┃ │ 4
    62 ┃ justResult Nothing = H.failWith Nothing "Expecting Just, got nothing"
       ┃ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       ┃ │ Expecting Just, got nothing
    
       ┏━━ test/Test/CLI/TextEnvelope/Golden/Genesis/Create.hs ━━━
    64 ┃ result :: Show a => a -> H.PropertyT IO a
    65 ┃ result a = H.forAll $ pure a
       ┃ │ 10000000
    
        ┏━━ test/Test/CLI/TextEnvelope/Golden/Genesis/Create.hs ━━━
     67 ┃ golden_genesisCreate :: Property
     68 ┃ golden_genesisCreate = OP.propertyOnce $ do
     69 ┃   liftIO $ IO.createDirectoryIfMissing True "tmp"
     70 ┃   tempDir <- liftIO $ IO.createTempDirectory "tmp" "test"
     71 ┃   let genesisFile = tempDir <> "/genesis.json"
     72 ┃   let cleanupPaths = [tempDir]
     73 ┃   
     74 ┃   fmtStartTime <- fmap formatIso8601 $ liftIO DT.getCurrentTime
     75 ┃ 
     76 ┃   (supply       , fmtSupply       ) <- fmap (withSnd show) $ forAll $ G.int (R.linear 10000000 4000000000)
        ┃   │ 10000000
     77 ┃   (delegateCount, fmtDelegateCount) <- fmap (withSnd show) $ forAll $ G.int (R.linear 4 19)
        ┃   │ 4
     78 ┃   (utxoCount    , fmtUtxoCount    ) <- fmap (withSnd show) $ forAll $ G.int (R.linear 4 19)
        ┃   │ 4
     79 ┃ 
     80 ┃   -- Create the genesis json file and required keys
     81 ┃   OP.execCardanoCLIParser cleanupPaths $
     82 ┃     OP.evalCardanoCLIParser
     83 ┃       [ "shelley","genesis","create"
     84 ┃       , "--testnet-magic", "12"
     85 ┃       , "--start-time", fmtStartTime
     86 ┃       , "--supply", fmtSupply
     87 ┃       , "--gen-genesis-keys", fmtDelegateCount
     88 ┃       , "--gen-utxo-keys", fmtUtxoCount
     89 ┃       , "--genesis-dir", tempDir
     90 ┃       ]
     91 ┃ 
     92 ┃   OP.assertFilesExist [genesisFile]
     93 ┃ 
     94 ┃   genesisContents <- liftIO $ IO.readFile genesisFile
     95 ┃ 
     96 ┃   H.annotate genesisContents
        ┃   │ {
        ┃   │     "activeSlotsCoeff": 5.0e-2,
        ┃   │     "protocolParams": {
        ┃   │         "poolDeposit": 0,
        ┃   │         "protocolVersion": {
        ┃   │             "minor": 0,
        ┃   │             "major": 0
        ┃   │         },
        ┃   │         "minUTxOValue": 0,
        ┃   │         "decentralisationParam": 1,
        ┃   │         "maxTxSize": 16384,
        ┃   │         "minPoolCost": 0,
        ┃   │         "minFeeA": 0,
        ┃   │         "maxBlockBodySize": 65536,
        ┃   │         "minFeeB": 0,
        ┃   │         "eMax": 18,
        ┃   │         "extraEntropy": {
        ┃   │             "tag": "NeutralNonce"
        ┃   │         },
        ┃   │         "maxBlockHeaderSize": 1100,
        ┃   │         "keyDeposit": 0,
        ┃   │         "nOpt": 100,
        ┃   │         "rho": 0,
        ┃   │         "tau": 0,
        ┃   │         "a0": 0
        ┃   │     },
        ┃   │     "protocolMagicId": 42,
        ┃   │     "genDelegs": {
        ┃   │         "8186a1e7a631f2edf7b83d8dce1ac896002730f5df3848e3d38182bc": {
        ┃   │             "delegate": "0bbacc100fd48ea617b5bb08f3ae537ef6447a0ee27292569283f947",
        ┃   │             "vrf": "504d3dbfda4636ce2037e4b6339242a8c29371d8c127256d66bd5dcfc73f8614"
        ┃   │         },
        ┃   │         "b302b767ab52d43e93236f784630fe584b5223444b0533c9e6aee700": {
        ┃   │             "delegate": "f3eddfac20e7277041c3abc6f93530034c291ce69f7c886e3aa4f0de",
        ┃   │             "vrf": "585aae3d459943444ef598395b2358708e2c15e66b7f3aae0a53051348a45ce3"
        ┃   │         },
        ┃   │         "a5cb234fff546e7b7c747c9809847b9cf81046bdf13c6b9603320387": {
        ┃   │             "delegate": "0b8d0ac264bc5a198fb7662e4c7d897a904adbe141acc7dcf0d30d15",
        ┃   │             "vrf": "3db55d09e8d94f7cdcfaf1ee7bbb6da51ec7b39d3aa2971b5b73ab766445d4a7"
        ┃   │         },
        ┃   │         "11265896a0937e5e0fd0844c4d7276c9f24b8d22b838753fd4b7ee48": {
        ┃   │             "delegate": "f40fc1dd9aed774b828852c365647dfbea1d346f2b10001e3a569ab3",
        ┃   │             "vrf": "1755b3228ce2b85ed5ab570534efc6bbf1cc152693a256c60a898afdfde5242a"
        ┃   │         }
        ┃   │     },
        ┃   │     "updateQuorum": 5,
        ┃   │     "networkId": "Testnet",
        ┃   │     "initialFunds": {
        ┃   │         "606e99b55a25f099d33cc48cdeaa665149ae10f021a1936c639abb09f5": 2500000,
        ┃   │         "608216da547bf781294c340fa4dda94161a8abdf46867821176a179c17": 2500000,
        ┃   │         "609f88eedbfdd5a78ddee853741a3f5454832243a47dcd9d11d62f938a": 2500000,
        ┃   │         "602b70034272c058760f736212267a26f47238a3f21c4e3a1b14597202": 2500000
        ┃   │     },
        ┃   │     "maxLovelaceSupply": 10000000,
        ┃   │     "networkMagic": 42,
        ┃   │     "epochLength": 432000,
        ┃   │     "staking": {
        ┃   │         "pools": {},
        ┃   │         "stake": {}
        ┃   │     },
        ┃   │     "systemStart": "2020-07-10T07:44:17Z",
        ┃   │     "slotsPerKESPeriod": 129600,
        ┃   │     "slotLength": 1,
        ┃   │     "maxKESEvolutions": 60,
        ┃   │     "securityParam": 2160
        ┃   │ }
     97 ┃ 
     98 ┃   actualSupply        <- justResult $ genesisContents ^? J.key "maxLovelaceSupply" . J._Integral
     99 ┃   actualStartTime     <- justResult $ genesisContents ^? J.key "systemStart" . J._String <&> T.unpack
    100 ┃   actualDelegateCount <- justResult $ genesisContents ^? J.key "genDelegs" . J._Object <&> HMS.size
    101 ┃   actualTotalSupply   <- result $ sum  $ genesisContents ^.. J.key "initialFunds" . J._Object . CL.to HMS.toList . each . CL._2 . J._Integral
    102 ┃   actualDelegates     <- justResult @J.Array Nothing -- $ genesisContents ^? J.key "genDelegs" . CL.to objectToArray . J._Array
    103 ┃ 
    104 ┃   actualSupply        === supply
    105 ┃   actualStartTime     === fmtStartTime
    106 ┃   actualDelegateCount === delegateCount
    107 ┃   actualDelegateCount === utxoCount
    108 ┃   actualTotalSupply   === supply        -- Check that the sum of the initial fund amounts matches the total supply
    109 ┃ 
    110 ┃   -- Check uniqueness and count of hash keys
    111 ┃   let hashKeys = actualDelegates ^.. each . J.nth 0 . J._String
    112 ┃ 
    113 ┃   S.size (S.fromList hashKeys) === length hashKeys -- This isn't strictless necessary because we use aeson which guarantees uniqueness of keys
    114 ┃   S.size (S.fromList hashKeys) === delegateCount
    115 ┃ 
    116 ┃   -- Check uniqueness and count of hash keys
    117 ┃   let delegateKeys = actualDelegates ^.. each . J.nth 1 . J.key "delegate" . J._String
    118 ┃ 
    119 ┃   S.size (S.fromList delegateKeys) === length delegateKeys
    120 ┃   S.size (S.fromList delegateKeys) === delegateCount
    121 ┃ 
    122 ┃   for_ [1 .. delegateCount] $ \i -> do
    123 ┃     -- Check Genesis keys
    124 ┃     assertFileOccurences 1 "Genesis signing key"      $ tempDir <> "/genesis-keys/genesis" <> show i <> ".skey"
    125 ┃     assertFileOccurences 1 "Genesis verification key" $ tempDir <> "/genesis-keys/genesis" <> show i <> ".vkey"
    126 ┃ 
    127 ┃     -- Check delegate keys
    128 ┃     assertFileOccurences 1 "Node operator signing key"                  $ tempDir <> "/delegate-keys/delegate" <> show i <> ".skey"
    129 ┃     assertFileOccurences 1 "Node operator verification key"             $ tempDir <> "/delegate-keys/delegate" <> show i <> ".vkey"
    130 ┃     assertFileOccurences 1 "Node operational certificate issue counter" $ tempDir <> "/delegate-keys/delegate" <> show i <> ".counter"
    131 ┃ 
    132 ┃     -- Check utxo keys
    133 ┃     assertFileOccurences 1 "Genesis UTxO signing key"       $ tempDir <> "/utxo-keys/utxo" <> show i <> ".skey"
    134 ┃     assertFileOccurences 1 "Genesis UTxO verification key"  $ tempDir <> "/utxo-keys/utxo" <> show i <> ".vkey"
  
    This failure can be reproduced by running:
    > recheck (Size 0) (Seed 6097814995919492232 16442572547884707391) golden_genesisCreate
  
  ✗ 1 failed.
```